tikv/raftstore/fsm_store.go

```diff
@@ -698,58 +723,357 @@ func (bs *raftBatchSystem) shutDown() {
 }
 
 func createRaftBatchSystem(cfg *Config) (*router, *raftBatchSystem) {
-	return nil, nil // TODO: stub
+	storeSender, storeFsm := newStoreFsm(cfg)
+	applyRouter, applySystem := createApplyBatchSystem(cfg)
+	router, system := createBatchSystem(int(cfg.StorePoolSize), int(cfg.StoreMaxBatchSize), storeSender, storeFsm)
+	raftBatchSystem := &raftBatchSystem{
+		system:      system,
+		applyRouter: applyRouter,
+		applySystem: applySystem,
+		router:      router,
+	}
+	return router, raftBatchSystem
 }
 
 /// Checks if the message is targeting a stale peer.
 ///
 /// Returns true means the message can be dropped silently.
 func (d *storeFsmDelegate) checkMsg(msg *rspb.RaftMessage) (bool, error) {
-	return false, nil // TODO: stub
+	regionID := msg.GetRegionId()
+	fromEpoch := msg.GetRegionEpoch()
+	msgType := msg.Message.MsgType
+	isVoteMsg := isVoteMessage(msg.Message)
+	fromStoreID := msg.FromPeer.StoreId
+
+	// Check if the target is tombstone,
+	stateKey := RegionStateKey(regionID)
+	localState := new(rspb.RegionLocalState)
+	err := getMsg(d.ctx.engine.kv, stateKey, localState)
+	if err != nil {
+		if err == badger.ErrKeyNotFound {
+			return false, nil
+		}
+		return false, err
+	}
+	if localState.State != rspb.PeerState_Tombstone {
+		// Maybe split, but not registered yet.
+		if isFirstVoteMessage(msg.Message) {
+			d.ctx.storeMetaLock.Lock()
+			defer d.ctx.storeMetaLock.Unlock()
+			meta := d.ctx.storeMeta
+			// Last check on whether target peer is created, otherwise, the
+			// vote message will never be comsumed.
+			if _, ok := meta.regions[regionID]; ok {
+				return false, nil
+			}
+			meta.pendingVotes = append(meta.pendingVotes, msg)
+			log.Infof("region %d doesn't exist yet, wait for it to be split.", regionID)
+			return true, nil
+		}
+		return false, errors.Errorf("region %d not exists but not tombstone: %s", regionID, localState)
+	}
+	log.Debugf("region %d in tombstone state: %s", regionID, localState)
+	region := localState.Region
+	regionEpoch := region.RegionEpoch
+	if localState.MergeState != nil {
+		// TODO: Merge
+		return true, nil
+	}
+	// The region in this peer is already destroyed
+	if IsEpochStale(fromEpoch, regionEpoch) {
+		log.Infof("tombstone peer receives a stale message. region_id:%d, from_region_epoch:%s, current_region_epoch:%s, msg_type:%s",
+			regionID, fromEpoch, regionEpoch, msgType)
+		notExist := findPeer(region, fromStoreID) == nil
+		d.ctx.handleStaleMsg(msg, regionEpoch, isVoteMsg && notExist, nil)
+		return true, nil
+	}
+	if fromEpoch.Version == regionEpoch.Version {
+		return false, errors.Errorf("tombstone peer [epoch: %s] received an invalid message %s, ignore it",
+			regionEpoch, msgType)
+	}
+	return false, nil
 }
 
 func (d *storeFsmDelegate) onRaftMessage(msg *rspb.RaftMessage) error {
-	return nil // TODO: stub
+	regionID := msg.RegionId
+	if err := d.ctx.router.send(regionID, Msg{Type: MsgTypeRaftMessage, Data: msg}); err != nil {
```

Comment:
How about channel is closed when shutdown?

Reply:
My mistake.